### PR TITLE
basic: port various pidfd/pidref helpers to PIDFD_GET_INFO and PIDFD_GET_*_NAMESPACE

### PR DESCRIPTION
**Pull request by YHNdnzj**
Supersedes #35308 (cherry-picked one commit and replaced the rest)

(I left a few comments that's folded by GitHub. Please make sure to check them too.)

### CODE REVIEW

**Comment by bluca** [comment]
> @bluca Does PIDFD_GET_INFO provide a way to distinguish pid from another pidns? I can't seem to find anything.

There's another existing ioctl to get the namespace info from a pidfd, `PIDFD_GET_PID_NAMESPACE`

**Comment by YHNdnzj** [comment]
> > @bluca Does PIDFD_GET_INFO provide a way to distinguish pid from another pidns? I can't seem to find anything.
> 
> There's another existing ioctl to get the namespace info from a pidfd, `PIDFD_GET_PID_NAMESPACE`

Yeah, but that would be one more ioctl() and extra inode_same_at() check. I wish this to be a flag of PIDFD_GET_INFO.

**Comment by bluca** [comment]
we've kept them separate instead, shouldn't be a problem doing more ioctls, there will be even more at some point for the security labels

File: src/basic/cgroup-util.c [hunk 1]
```diff
@@ -72,6 +72,28 @@ int cg_cgroupid_open(int cgroupfs_fd, uint64_t id) {
...
+        if (cgfd < 0)
+                return cgfd;
+
+        _cleanup_free_ char *path = NULL;
+
+        r = fd_get_path(cgfd, &path);
+        if (r < 0)
+                return r;
+
+        if (isempty(path_startswith(path, "/sys/fs/cgroup/")))
```
Comment (poettering) [inline src/basic/cgroup-util.c]: hmm, for the root cgroup, this is going to cause failure, no?

File: src/nsresourced/userns-registry.c [hunk 2]
```diff
@@ -525,42 +525,20 @@ int userns_info_add_cgroup(UserNamespaceInfo *userns, uint64_t cgroup_id) {
...
-        _cleanup_free_ char *path = NULL;
-        r = fd_get_path(cgroup_fd, &path);
         if (r < 0)
                 return log_debug_errno(r, "Failed to get path of cgroup %" PRIu64 ", ignoring: %m", cgroup_id);
 
-        const char *e = path_startswith(path, "/sys/fs/cgroup/");
-        if (!e)
-                return log_debug_errno(SYNTHETIC_ERRNO(EPERM), "Got cgroup path that doesn't start with /sys/fs/cgroup/, refusing: %s", path);
-        if (isempty(e))
-                return log_debug_errno(SYNTHETIC_ERRNO(EPERM), "Got root cgroup path, which can't be right, refusing.");
```
Comment (poettering) [inline src/nsresourced/userns-registry.c]: this shoudl really be kept. for the userns stuff this check matters, but not for the generic cg_path_from_cgroupid

File: src/nsresourced/userns-registry.c [hunk 3]
```diff
@@ -525,42 +525,20 @@ int userns_info_add_cgroup(UserNamespaceInfo *userns, uint64_t cgroup_id) {
...
-
-        parent_fd = openat(cgroup_fd, "..", O_CLOEXEC|O_DIRECTORY);
-        if (parent_fd < 0)
-                return log_debug_errno(errno, "Failed to open parent cgroup of %" PRIu64 ", ignoring: %m", cgroup_id);
-
-        cgroup_fd = safe_close(cgroup_fd);
+        log_debug("Destroying cgroup %" PRIu64 " (%s)", cgroup_id, path);
 
-        r = rm_rf_child(parent_fd, fname, REMOVE_ONLY_DIRECTORIES|REMOVE_CHMOD);
+        r = rm_rf(path, REMOVE_ROOT|REMOVE_ONLY_DIRECTORIES|REMOVE_CHMOD);
```
Comment (poettering) [inline src/nsresourced/userns-registry.c]: i don't think that's an improvement. if we already have the fd we really should go by fd

**Comment by poettering** [comment]
I guess pidref_get_uid() could be ported over too, pretty easily.

maybe pidref_is_my_child() too

File: src/basic/pidfd-util.c [hunk 4]
```diff
@@ -12,6 +13,31 @@
 #include "pidfd-util.h"
 #include "string-util.h"
 
+int pidfd_get_namespace(int fd, unsigned int ns_type_flag) {
```
Comment (yuwata) [inline src/basic/pidfd-util.c]: `unsigned int` -> `unsigned`
or
`unsigned int` -> `unsigned long` ?

**Comment by YHNdnzj** [comment]
Hrm, open_by_handle_at() + cg_file_handle requires privilege, and even when privileged we cannot handle remote cgroup properly. I'd tentatively drop the commits that port cg_pidref_get_path() to pidfd helpers, and figure out how to do this / if it's viable later.

cc @poettering

File: src/basic/namespace-util.h [hunk 5]
```diff
@@ -24,11 +24,15 @@ extern const struct namespace_info {
         const char *proc_name;
         const char *proc_path;
         unsigned int clone_flag;
+        unsigned int pidfd_get_ns_flag;
```
Comment (yuwata) [inline src/basic/namespace-util.h]: `unsigned int` -> `unsigned` or `unsigned long`

File: src/basic/pidfd-util.c [hunk 6]
```diff
@@ -145,3 +147,26 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
...
         return current_pid != pid ? -ESRCH : 0;
 }
+
+int pidfd_get_inode_id(int fd, uint64_t *ret) {
+        static int cached_supported = -1;
+
+        assert(fd >= 0);
+
+        if (cached_supported < 0) {
+                cached_supported = fd_is_fs_type(fd, PID_FS_MAGIC);
```
Comment (yuwata) [inline src/basic/pidfd-util.c]: Hm? Why cached?? The result depends on the input fd, no?

Please add simple unit tests for the function.

File: src/basic/cgroup-util.c [hunk 7]
```diff
@@ -68,6 +68,58 @@ int cg_cgroupid_open(int cgroupfs_fd, uint64_t id) {
...
+        _cleanup_close_ int cgfd = -EBADF;
+        int r;
+
+        cgfd = cg_cgroupid_open(cgroupfs_fd, id);
+        if (cgfd < 0)
+                return cgfd;
+
+        _cleanup_free_ char *path = NULL;
+
+        r = fd_get_path(cgfd, &path);
```
Comment (yuwata) [inline src/basic/cgroup-util.c]: BTW, IIRC, we do not have any rules about variable declaration in a middle of function, but I am not a fan of inserting an empty line between the declaration and its use. At least I and Lennart do not insert such.

File: src/basic/cgroup-util.c [hunk 8]
```diff
@@ -68,6 +68,58 @@ int cg_cgroupid_open(int cgroupfs_fd, uint64_t id) {
...
+        /* This is cgroupfs so we know the size of the handle, thus no need to loop around like
+         * name_to_handle_at_loop() does in mountpoint-util.c */
+        if (name_to_handle_at(AT_FDCWD, path, &fh.file_handle, &mnt_id, 0) < 0)
+                return -errno;
+
+        *ret = CG_FILE_HANDLE_CGROUPID(fh);
+        return 0;
+}
+
+int cg_fd_get_cgroupid(int fd, uint64_t *ret) {
```
Comment (yuwata) [inline src/basic/cgroup-util.c]: Maybe merge the two into `cg_get_cgroup_id_at()` or so? And make them inline functions?

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/basic/pidfd-util.c (hunk 6 above)
Comment (YHNdnzj) [reply src/basic/pidfd-util.c]: We only expect pidfds, but initially they were implemented as anonymous inodes. Kernel v6.9 introduced a pseudo pidfs and inode number is meaningful only there.

File: src/basic/cgroup-util.c (hunk 7 above)
Comment (YHNdnzj) [reply src/basic/cgroup-util.c]: Dunno. Unless the initialization is a function call, I really dislike placing stuff tightly together...

File: src/basic/pidfd-util.c (hunk 6 above)
Comment (yuwata) [reply src/basic/pidfd-util.c]: Then, that's not a good idea to merge this with `pidfd_inode_ids_supported()`.
When this is exposed, an arbitrary fd may be passed mistakenly, and we cannot set any assertion for the misuse.
So, keep `pidfd_inode_ids_supported()` as is, and call it in this function to check if the functionality is supported. The result is cached. Hence should not impact performance.

File: src/basic/cgroup-util.c (hunk 7 above)
Comment (yuwata) [reply src/basic/cgroup-util.c]: Not a matter in this PR. But I really dislike the way. Maybe, let's discuss somewhere later and update our coding style doc.

File: src/basic/pidfd-util.c (hunk 6 above)
Comment (YHNdnzj) [reply src/basic/pidfd-util.c]: >  When this is exposed, an arbitrary fd may be passed mistakenly, and we cannot set any assertion for the misuse.

But what this function does is essentially getting inode number. So we really cannot prevent such case (unless `fd_is_fs_type()` is checked for each and every invocation. `ioctl`s, similarly, can mean completely different stuff on different fds. We generally do not protect against totally non-sensical mistakes like this?

Comment (yuwata) [reply src/basic/pidfd-util.c]: Even though, it is not necessary to introduce any further pitfall.
The problem here is, if this is called first time with a wrong fd, then the result is cached, and the inode number cannot be used by any later calls. And, it is hard to notice the misuse here. We can avoid such misuse here, then why not to do?

Moreover, I'd like to check if the passed fd a pidfd before calling ioctl()s in other places. That can be done if `PID_FS_MAGIC` is supported.

Comment (YHNdnzj) [reply src/basic/pidfd-util.c]: All these seem too much churn than it's worth... We have limited call sites, just like all other uses of ioctl. And this isn't particularly cheap. Just like we assert on `fd >= 0`, not `fd_validate()`.

Comment (yuwata) [reply src/basic/pidfd-util.c]: At least, keeping the check as it was here does not have any performance impact.

Safe guard before ioctl() may be too much, and slightly off topic. Feel free to ignore.

File: src/basic/pidfd-util.c [hunk 9]
```diff
@@ -7,11 +7,32 @@
...
 #include "memory-util.h"
+#include "missing_magic.h"
 #include "parse-util.h"
 #include "path-util.h"
 #include "pidfd-util.h"
+#include "process-util.h"
+#include "stat-util.h"
 #include "string-util.h"
 
+static int is_pidfs = -1;
```
Comment (poettering) [inline src/basic/pidfd-util.c]: maybe call this `have_pidfs`?

File: src/basic/pidfd-util.c [hunk 10]
```diff
@@ -64,3 +85,24 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
...
+        assert(fd >= 0);
+
+        r = pidfd_check_pidfs();
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return -EOPNOTSUPP;
+
+        struct stat st;
+
```
Comment (poettering) [inline src/basic/pidfd-util.c]: i'd drop this line

File: src/basic/errno-util.h [hunk 11]
```diff
@@ -172,6 +172,12 @@ static inline bool ERRNO_IS_NEG_NOT_SUPPORTED(intmax_t r) {
 }
 _DEFINE_ABS_WRAPPER(NOT_SUPPORTED);
 
+/* ioctl() with unsupported command/arg might additionally return EINVAL */
+static inline bool ERRNO_IS_NEG_IOCTL_NOT_SUPPORTED(intmax_t r) {
+        return ERRNO_IS_NEG_NOT_SUPPORTED(r) || r == -EINVAL;
```
Comment (poettering) [inline src/basic/errno-util.h:177]: ah, nice idea, i think i like it

File: src/basic/pidfd-util.c [hunk 12]
```diff
@@ -33,6 +34,34 @@ static int pidfd_check_pidfs(void) {
...
+        assert(fd >= 0);
+
+        /* If we know ahead of time that pidfs is unavailable, shortcut things. But otherwise we don't
+         * call pidfd_check_pidfs() here, which is kinda extraneous and our own cache is required
+         * anyways (pidfs is introduced in kernel 6.9 while ioctl support there is added in 6.11). */
+        if (is_pidfs == 0 || !cached_supported)
+                return -EOPNOTSUPP;
+
+        int nsfd = ioctl(fd, ns_type_flag);
```
Comment (poettering) [inline src/basic/pidfd-util.c]: hmm, ns_type_flag is an ioctl command number? shouldn't be called "flag" then.

reading through this PR it's not clear to me with what kind of 2nd params this function shall be called, this deserves a comment.

(i see now this is mentioned in the commit msg, but this should be in the comment too)

File: src/basic/pidfd-util.c [hunk 13]
```diff
@@ -33,6 +34,34 @@ static int pidfd_check_pidfs(void) {
...
+        if (nsfd < 0) {
+                /* ERRNO_IS_(IOCTL_)NOT_SUPPORTED cannot be used here, because kernel returns -EOPNOTSUPP
+                 * if the NS is disabled at build time. */
```
Comment (poettering) [inline src/basic/pidfd-util.c]: i think this should be reworded a bit, i.e. it should emphasize that we need to look at the precise errors to be able to distinguish "ioctl() API not supported" from "namespace type not supported"

File: src/basic/pidfd-util.c [hunk 14]
```diff
@@ -33,6 +34,34 @@ static int pidfd_check_pidfs(void) {
...
+                if (IN_SET(errno, ENOTTY, EINVAL)) {
+                        cached_supported = false;
+                        return -EOPNOTSUPP;
+                }
+                if (errno == EOPNOTSUPP) /* Translate to something more distinguishable */
+                        return -ENOPKG;
```
Comment (poettering) [inline src/basic/pidfd-util.c:63]: can you summarize the expected error returns in a comment at the top?

File: src/basic/namespace-util.c [hunk 15]
```diff
@@ -191,50 +191,49 @@ int namespace_enter(int pidns_fd, int mntns_fd, int netns_fd, int userns_fd, int
...
+        /* Checks whether the specified file descriptor refers to a namespace (of type if type != _NAMESPACE_INVALID). */
 
-        if (fstatfs(fd, &s) < 0)
-                return -errno;
+        assert(fd >= 0);
+        assert(type < _NAMESPACE_TYPE_MAX);
 
-        if (!is_fs_type(&s, NSFS_MAGIC)) {
-                /* On really old kernels, there was no "nsfs", and network namespace sockets belonged to procfs
-                 * instead. Handle that in a somewhat smart way. */
```
Comment (poettering) [inline src/basic/namespace-util.c:207]: hmm, if this is dropped we'll update the baseline to 4.11 or so? did we all agree on that? (don't remember the last discussions). If so, this PR should also update the README accordingly.

File: src/basic/namespace-util.c [hunk 16]
```diff
@@ -191,50 +191,49 @@ int namespace_enter(int pidns_fd, int mntns_fd, int netns_fd, int userns_fd, int
...
 
-                return -errno;
-        }
+        r = fd_is_namespace(fd, type);
+        if (r < 0)
+                return r;
+        if (r == 0) /* Not a namespace or not of the right type? */
+                return -EUCLEAN;
+
+        _cleanup_close_ int our_ns = namespace_open_by_type(type);
```
Comment (poettering) [inline src/basic/namespace-util.c:284]: this will be a tiny bit slower of course, since you open an then fstat() the thing, while previously we directly did stat() on the path.

I figure this is fine, but did you do this on purpose?

File: src/basic/namespace-util.h [hunk 17]
```diff
@@ -23,12 +23,16 @@ enum NamespaceType {
 extern const struct namespace_info {
         const char *proc_name;
         const char *proc_path;
-        unsigned int clone_flag;
+        unsigned long clone_flag;
+        unsigned long pidfd_get_ns_flag;
```
Comment (poettering) [inline src/basic/namespace-util.h]: this is not a flag, but an ioctl command number

File: src/basic/process-util.c [hunk 18]
```diff
@@ -1199,12 +1199,12 @@ int pid_from_same_root_fs(pid_t pid) {
 }
 
 bool is_main_thread(void) {
-        static thread_local int cached = 0;
+        static thread_local int cached = -1;
 
-        if (_unlikely_(cached == 0))
-                cached = getpid_cached() == gettid() ? 1 : -1;
+        if (_unlikely_(cached < 0))
```
Comment (poettering) [inline src/basic/process-util.c]: might as well drop the `_unlikely_()` here, afaics there's a good chance we call this function once only in each tool, which hence means this optimization won't help but likely make things worse...

File: src/basic/pidfd-util.c [hunk 19]
```diff
@@ -170,6 +170,60 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
         return current_pid != pid ? -ESRCH : 0;
 }
 
+int pidfd_get_ppid(int fd, pid_t *ret) {
+        struct pidfd_info info = { .mask = PIDFD_INFO_PID };
```
Comment (poettering) [inline src/basic/pidfd-util.c:180]: doesn't matter, but i'd move this down, immediately before the pidfd_get_info() where you need it.

File: src/basic/process-util.c [hunk 20]
```diff
@@ -582,6 +582,14 @@ int pidref_get_uid(const PidRef *pid, uid_t *ret) {
...
                 return -EREMOTE;
 
+        if (pid->fd >= 0) {
+                r = pidfd_get_uid(pid->fd, ret);
+                if (r != -EOPNOTSUPP)
+                        return r;
+        }
+
+        uid_t uid;
+
```
Comment (poettering) [inline src/basic/process-util.c]: i'd remove this empty line, i.e. always declare variables that are needed for some immediately following functional call actually immediately before it.

File: src/basic/process-util.c [hunk 21]
```diff
@@ -722,6 +730,36 @@ int get_process_ppid(pid_t pid, pid_t *ret) {
...
+                return -EREMOTE;
+
+        if (pidref->fd >= 0) {
+                r = pidfd_get_ppid(pidref->fd, ret);
+                if (r != -EOPNOTSUPP)
+                        return r;
+        }
+
+        pid_t ppid;
+
```
Comment (poettering) [inline src/basic/process-util.c]: as above

File: src/nsresourced/userns-registry.c [hunk 22]
```diff
@@ -558,16 +558,9 @@ static int userns_destroy_cgroup(uint64_t cgroup_id) {
         if (parent_fd < 0)
                 return log_debug_errno(errno, "Failed to open parent cgroup of %" PRIu64 ", ignoring: %m", cgroup_id);
 
-        /* Safety check, never leave cgroupfs */
-        r = fd_is_fs_type(parent_fd, CGROUP2_SUPER_MAGIC);
-        if (r < 0)
-                return log_debug_errno(r, "Failed to determine if parent directory of cgroup %" PRIu64 " is still a cgroup, ignoring: %m", cgroup_id);
-        if (!r)
-                return log_debug_errno(SYNTHETIC_ERRNO(EPERM), "Parent directory of cgroup %" PRIu64 " is not a cgroup, refusing.", cgroup_id);
```
Comment (poettering) [inline src/nsresourced/userns-registry.c]: i think this check should be kept. after all the rm_rf() logic would happily erase a tmpfs here, which we don't want here, it really should be a cgroupfs and nothing else.

File: src/basic/cgroup-util.c [hunk 23]
```diff
@@ -68,6 +68,28 @@ int cg_cgroupid_open(int cgroupfs_fd, uint64_t id) {
...
+int cg_path_from_cgroupid(int cgroupfs_fd, uint64_t id, char **ret) {
+        _cleanup_close_ int cgfd = -EBADF;
+        int r;
+
+        cgfd = cg_cgroupid_open(cgroupfs_fd, id);
+        if (cgfd < 0)
+                return cgfd;
+
+        _cleanup_free_ char *path = NULL;
+
```
Comment (poettering) [inline src/basic/cgroup-util.c]: as above

File: src/basic/cgroup-util.c [hunk 24]
```diff
@@ -860,6 +860,10 @@ int cg_pidref_get_path(const char *controller, const PidRef *pidref, char **ret_
         if (!pidref_is_set(pidref))
                 return -ESRCH;
 
+        // XXX: Ideally we'd use pidfd_get_cgroupid() + cg_path_from_cgroupid() here, to extract this
+        // bit of information from pidfd directly. However, the latter requires privilege and it's
+        // not entirely clear how to handle cgroups from outer namespace.
```
Comment (poettering) [inline src/basic/cgroup-util.c:866]: outer → out of?

Comment (poettering) [reply src/basic/cgroup-util.c:866]: we probably could try cg_path_from_cgroupid() and then fall back if it fails with EPERM.

i think the path for cgroups from outside of the namespace are marked specially, start with `../` or so something like that, so that one case recognize them. don't remember the details, but this was a solvable problem iirc

Comment (YHNdnzj) [reply src/basic/cgroup-util.c:866]: I tried, and the path attached to fd is simply `/`...

File: src/basic/process-util.c (hunk 18 above)
Comment (YHNdnzj) [reply src/basic/process-util.c]: -> #35561

File: src/nsresourced/userns-registry.c (hunk 22 above)
Comment (YHNdnzj) [reply src/nsresourced/userns-registry.c]: -> #35563

File: src/basic/namespace-util.c (hunk 15 above)
Comment (YHNdnzj) [reply src/basic/namespace-util.c:207]: We agreed to bump the minimum to v5.4 too in 2025, but I don't want to do that in this PR. The PR that removes cgroup v1 is probably a better place IMO (that's also a wip from my side).

File: src/basic/namespace-util.c (hunk 16 above)
Comment (YHNdnzj) [reply src/basic/namespace-util.c:284]: We could theoretically be _way_ slower given that we prefer pidfd_open() first, i.e. avoiding procfs. But this is not a hot path IMO, and shouldn't really have any visible impact.

File: src/basic/pidfd-util.c (hunk 19 above)
Comment (YHNdnzj) [reply src/basic/pidfd-util.c:180]: I prefer the current style better. The function is small and the var is used all across it after all.

File: src/basic/pidref.c [hunk 25]
```diff
@@ -46,19 +27,14 @@ int pidref_acquire_pidfd_id(PidRef *pidref) {
         if (pidref->fd_id > 0)
                 return 0;
 
-        r = pidfd_inode_ids_supported();
-        if (r < 0)
+        r = pidfd_get_inode_id(pidref->fd, &pidref->fd_id);
+        if (r < 0) {
+                if (r != -EOPNOTSUPP)
```
Comment (poettering) [inline src/basic/pidref.c]: doesn't matter, but iÄd probably use ERRNO_IS_NET_NOT_SUPPORTED here, just because

File: src/basic/namespace-util.c (hunk 15 above)
Comment (poettering) [reply src/basic/namespace-util.c:207]: i'd update README more aggressively tbh, even if its updated many times in this PR cycles. It just makes things easier to debug later on just by looking at git history

File: src/basic/namespace-util.c [hunk 26]
```diff
@@ -48,6 +49,62 @@ NamespaceType clone_flag_to_namespace_type(unsigned long clone_flag) {
...
+
+        assert(pidref_is_set(pidref));
+        assert(type >= 0 && type < _NAMESPACE_TYPE_MAX);
+
+        if (pidref_is_remote(pidref))
+                return -EREMOTE;
+
+        if (pidref->fd >= 0) {
+                r = pidfd_get_namespace(pidref->fd, namespace_info[type].pidfd_get_ns_ioctl_cmd);
+                if (r != -EOPNOTSUPP)
```
Comment (poettering) [inline src/basic/namespace-util.c]: given that you now return a recognizable ENOPKG error for the other kind of "not supported", you could just use ERRNO_IS_NOT_SUPPORTED() here, just to make this less surprising.

(or to say this differently: if we do not use ERRNO_IS_NOT_SUPPORTED() then we really should add a comment why not. But there's no reason to not use it here, hence we should)

File: src/basic/namespace-util.c [hunk 27]
```diff
@@ -48,6 +49,62 @@ NamespaceType clone_flag_to_namespace_type(unsigned long clone_flag) {
...
+                        return r;
+        }
+
+        if (need_verify) /* The caller shall call pidref_verify() later */
+                *need_verify = true;
```
Comment (poettering) [inline src/basic/namespace-util.c:68]: hmm, why is this not a regular return param? i.e. named ret_need_verify and only initialized on success?

File: src/basic/namespace-util.c [hunk 28]
```diff
@@ -66,53 +124,47 @@ int pidref_namespace_open(
...
 
         if (ret_userns_fd) {
-                const char *userns;
-
-                userns = pid_namespace_path(pidref->pid, NAMESPACE_USER);
-                userns_fd = open(userns, O_RDONLY|O_NOCTTY|O_CLOEXEC);
-                if (userns_fd < 0 && errno != ENOENT)
-                        return -errno;
+                userns_fd = pidref_namespace_open_by_type_internal(pidref, NAMESPACE_USER, &need_verify);
+                if (userns_fd < 0 && !IN_SET(userns_fd, -ENOENT, -ENOPKG))
```
Comment (poettering) [inline src/basic/namespace-util.c:146]: why only allowlist these two extra errors for userns? so far we said the other namespacing types are ptional too

File: src/basic/process-util.c [hunk 29]
```diff
@@ -1149,6 +1133,14 @@ int pidref_is_unwaited(const PidRef *pid) {
         return true;
 }
 
+int pid_is_unwaited(pid_t pid) {
+
+        if (pid == 0)
+                return false;
```
Comment (poettering) [inline src/basic/process-util.c]: return true, not false here, no?

File: src/basic/cgroup-util.c (hunk 24 above)
Comment (poettering) [reply src/basic/cgroup-util.c:866]: that's annoying, but we could recognize this case i guess, and then do the traditional fallback too

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/basic/namespace-util.c (hunk 28 above)
Comment (YHNdnzj) [reply src/basic/namespace-util.c:146]: Well, mountns and netns are definitely required, as documented in README. pidns is something everyone enable, so I think the current handling is reasonable. Also, it's not changed by this PR.

File: src/basic/namespace-util.c (hunk 27 above)
Comment (YHNdnzj) [reply src/basic/namespace-util.c:68]: Because the caller must initialize it to false, and if there's no need to verify we can't and won't reset it to false.

File: src/basic/cgroup-util.c (hunk 24 above)
Comment (YHNdnzj) [reply src/basic/cgroup-util.c:866]: Well, there has been enough churn that I deem the switch not worth it for now. I'd like to keep things as-is and move the discussion to another PR (I tried to make this work, but CI fails, and this shouldn't be blocked forever)

**Comment by YHNdnzj** [comment]
I left a few comments that's folded by GitHub. Please check them too.

File: src/basic/namespace-util.c (hunk 28 above)
Comment (poettering) [reply src/basic/namespace-util.c:146]: grmbl, i actually intended netns definitely and mountns too to be optional...

i see the comment in README about localed requiring netns, but frankly we probably should address that and add PrivateNetwork=try or so, in which case we gracefully degrade...

File: src/basic/namespace-util.c (hunk 27 above)
Comment (poettering) [reply src/basic/namespace-util.c:68]: comment might be nice

File: src/core/socket.c [hunk 30]
```diff
@@ -898,30 +901,35 @@ static int instance_from_socket(int fd, unsigned nr, char **instance) {
...
         case AF_UNIX: {
                 struct ucred ucred;
-                int k;
 
-                k = getpeercred(fd, &ucred);
-                if (k >= 0) {
-                        if (asprintf(&r,
-                                     "%u-"PID_FMT"-"UID_FMT,
-                                     nr, ucred.pid, ucred.uid) < 0)
+                r = getpeercred(fd, &ucred);
```
Comment (poettering) [inline src/core/socket.c:905]: once #35839 is merged this could probably be written in a shorter way by using getpeerpidfd() and pidref_acquire_pidfd_id()

File: src/basic/pidfd-util.c [hunk 31]
```diff
@@ -188,7 +188,7 @@ int pidfd_get_ppid(int fd, pid_t *ret) {
 
         assert(FLAGS_SET(info.mask, PIDFD_INFO_PID));
 
-        if (info.ppid == 0)
+        if (info.ppid == 0) /* See comments in pid_get_ppid() */
```
Comment (poettering) [inline src/basic/pidfd-util.c:191]: this got squashed into the wrong commit, should be one commit earlier i  guess

File: src/basic/process-util.c [hunk 32]
```diff
@@ -1046,41 +1082,32 @@ int getenv_for_pid(pid_t pid, const char *field, char **ret) {
...
-        if (pid <= 1)
+        if (pidref_is_remote(pid))
+                return -EREMOTE;
+
+        if (pid->pid == 1 || pidref_is_self(pid))
                 return false;
 
-        r = get_process_ppid(pid, &ppid);
+        pid_t ppid;
+        r = pidref_get_ppid(pid, &ppid);
```
Comment (poettering) [inline src/basic/process-util.c:1098]: unrelated to this series, but i figure we should return false here in case we see EADDRNOTAVAIL... a processes inserted from elsewhere is definitely not our child

**Comment by poettering** [comment]
(i'll fix a follow-up pr with the two relevant things i pointed out)

File: src/core/socket.c (hunk 30 above)
Comment (poettering) [reply src/core/socket.c:905]: i take this back, it doesn't get shorter. i tried and failed.
